### PR TITLE
Import iOS Starter Kit

### PR DESCRIPTION
**Pull request by marcisme**
Import the initial version of the iOS Starter Kit. This is an example of how to build a video app with the Daily iOS SDK using the latest `SwiftUI` features, with Xcode previews and some tests.

### CODE REVIEW

File: DailyKit/CallManager/CallManager.swift [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+    required init(callClient: CallClient) {
+        self.callClient = callClient
+        self.participantsBuilder = CallParticipants.Builder(
+            local: callClient.participants.local.asCallParticipant
+        )
+        self.subjects = CallClientSubjects(
+            url: callClient.url,
+            callState: callClient.callState,
+            camera: CallCamera(callClient.inputs.camera),
```
Comment (lazeratops) [inline DailyKit/CallManager/CallManager.swift:43]: Ok noob question about the property setup here: It looks like we're referencing properties from `callClient` to build the subjects, and then in this class making these call client properties available via:

1. The `callClient` constant (`private let callClient: CallClient`)
1. The `subjects` constant (`private let subjects: CallClientSubjects`)
1. Individual getters (?) that also reference `subject`s (e.g. `public var camera: CallCamera { subjects.camera.value }`)

I am guessing that last one might be to expose a public getter to whoever has a `CallManager` instance. Is that correct? Then I wonder about the other two - what does `subjects` provide that just referencing directly to `callClient` does not? Is its purpose just to hold the `publishers` collection? (Just want to make sure I wrap my head around this right)

Thanks!

File: DailyKit/CallManager/CallManager.swift [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        callClient.set(username: username)
+    }
+
+    public func join(url: URL) {
+        callClient.join(url: url)
+        subjects.url.send(url)
+
+        // TODO: Find a cleaner way to handle the joining transition.
+        // Send a `joining` value, so the UI will immediately transition to the waiting screen.
+        subjects.callState.send(.joining)
```
Comment (lazeratops) [inline DailyKit/CallManager/CallManager.swift:108]: How long does it take for the client's call state to update? Is it worth having a tiny bit of lag to keep the call client as the single source of truth? (I realize you already thought through this judging by your comment above, so this is more for my own curiosity)

File: DailyKit/CallManager/Models/CallCamera.swift [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    public let video: CallMuteState
+    public let mode: FacingMode
+
+    public init(video: CallMuteState, mode: FacingMode) {
+        self.video = video
+        self.mode = mode
+    }
+
+    public var isMuted: Bool { video == .muted }
```
Comment (lazeratops) [inline DailyKit/CallManager/Models/CallCamera.swift]: Super nit: would it be more intuitive for reading to have this var up with the costs above, before the constructor/initializer/whatever it is in Swift-land?

File: DailyKit/CallManager/Models/CallMuteState.swift [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    init(
+        _ settings: MicrophoneInputSettings,
+        status: AVAuthorizationStatus
+    ) {
+        switch (settings.isEnabled, status) {
+        case (true, .authorized):
+            self = .unmuted
+
+        case (_, _):
```
Comment (lazeratops) [inline DailyKit/CallManager/Models/CallMuteState.swift:45]: Is this like a default-case kind of thing, or what does it mean?

File: DailyKit/CallManager/Models/CallParticipant.swift [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        static let defaultUsername: String = "Guest"
+    }
+
+    public let id: AnyHashable
+    public internal(set) var username: String
+    public let isLocal: Bool
+    public let hasAudio: Bool
+    public let videoTrack: VideoTrack?
+    public var videoScaleMode: VideoView.VideoScaleMode { isLocal ? .fit : .fill }
+    public let hasVideo: Bool
```
Comment (lazeratops) [inline DailyKit/CallManager/Models/CallParticipant.swift:25]: Nit: might be nice to have `hasAudio` and `hasVideo` in closer proximity to each other for reading

File: DailyKit/CallManager/Models/CallParticipant.swift [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+extension CallParticipant {
+    init(_ participant: Participant) {
+        self.id = participant.id
+        self.username = participant.info.username ?? Constants.defaultUsername
+        self.isLocal = participant.info.isLocal
+        self.hasAudio = participant.media?.microphone.state == .playable
+        let videoTrack = participant.media.flatMap { media in
```
Comment (lazeratops) [inline DailyKit/CallManager/Models/CallParticipant.swift]: Nit: Since `participant.info` and `participant.media` are reused here, might be nice to have them in consts vs re-accessing.

File: DailyKit/CallManager/Models/CallParticipant.swift [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    public let isLocal: Bool
+    public let hasAudio: Bool
+    public let videoTrack: VideoTrack?
+    public var videoScaleMode: VideoView.VideoScaleMode { isLocal ? .fit : .fill }
+    public let hasVideo: Bool
+}
+
+extension CallParticipant {
+    init(_ participant: Participant) {
+        self.id = participant.id
```
Comment (lazeratops) [inline DailyKit/CallManager/Models/CallParticipant.swift:61]: Will this be the session ID or the user ID?

File: DailyKit/CallManager/Models/CallParticipants.swift [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+            joinOrder.append(participant.id)
+            joinOrder = Array(joinOrder.prefix(limit))
+
+            remote[participant.id] = participant
+        }
+
+        mutating func handleUpdated(_ participant: CallParticipant) {
+            if participant.isLocal {
+                local = participant
```
Comment (lazeratops) [inline DailyKit/CallManager/Models/CallParticipants.swift]: Wouldn't this be up to date from the `callClient` reference it was originally initialized with? Or is that a copy?

File: DailyKit/CallManager/Models/CallParticipants.swift [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+            assert(participant.isLocal == false, "The local participant cannot be removed.")
+
+            joinOrder.removeAll { $0 == participant.id }
+            speakerOrder.removeAll { $0 == participant.id }
+
+            remote.removeValue(forKey: participant.id)
+        }
+    }
+
+    public var count: Int { 1 + remote.count }
```
Comment (lazeratops) [inline DailyKit/CallManager/Models/CallParticipants.swift:6]: Nit: wondering if for readability it would be nicer to move these property definitions to the top of the class (but since I notice you sometimes put these in other locations in the class I'm starting to suspect Swift convention is different)

File: DailyKitTests/CallParticipantsTests.swift [hunk 10]
```diff
@@ -0,0 +1,219 @@
+@testable import DailyKit
+import XCTest
+
+final class CallParticipantsTests: XCTestCase, TestSupport {}
+
+// MARK: - Count
```
Comment (lazeratops) [inline DailyKitTests/CallParticipantsTests.swift]: Suggest rewording these comments slightly for clarity. E.g., removing name, maybe a little more info? "Test participant counts" or "Participant count tests" (or something)

File: DailyKitTests/CallParticipantsVisibleTests.swift [hunk 11]
```diff
@@ -0,0 +1,164 @@
+@testable import DailyKit
```
Comment (lazeratops) [inline DailyKitTests/CallParticipantsVisibleTests.swift]: I love seeing all these tests :D

File: DailyStarterKit/Layouts/GridLayoutView.swift [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+        }
+    }
+
+    @EnvironmentObject private var model: Model
+
+    @Environment(\.callLayout) private var layout: CallLayout
+
+    @Namespace private var namespace: Namespace.ID
+
+    private let animation: Animation = .default //.easeInOut(duration: 0.3)
```
Comment (lazeratops) [inline DailyStarterKit/Layouts/GridLayoutView.swift]: Reminder to remove comment if not needed

**Review by lazeratops — COMMENTED** [review]
I had a first read through and left mostly just a few clarifying questions for my own benefit and nits. 

Being new to Swift I look forward to seeing @regexident 's feedback as well, but from a non-Swift-specific perspective, to me this appears really well structured as a demo/starter kit. Easy to read and follow presuming some comments will be added (more on that below). 

Also love the collection of tests (which as a reader helped me get more insight into what the expected behavior of certain components is meant to be).

You mentioned you'd be going through and fleshing out the README and commenting, so I did not point out anything related to that here. I think the code could definitely benefit from more guiding comments throughout, as we're publishing it as a public example of how to build with Daily for people who may not already be familiar.

File: DailyKit/CallManager/Models/CallMuteState.swift [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+import AVFoundation
+
+public enum CallMuteState: Equatable {
+    case muted, unmuted
+}
+
+extension CallMuteState {
+    init(
+        _ settings: CameraInputSettings,
+        status: AVAuthorizationStatus
```
Comment (marcisme) [inline DailyKit/CallManager/Models/CallMuteState.swift:21]: Verify expected behavior for authorization status. We probably don't need this here, and we need to handle tapping the buttons when authorization has been denied.

File: DailyKit/CallManager/Models/CallParticipant.swift [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        self.username = participant.info.username ?? Constants.defaultUsername
+        self.isLocal = participant.info.isLocal
+        self.hasAudio = participant.media?.microphone.state == .playable
+        let videoTrack = participant.media.flatMap { media in
+            // Prefer screen video when available.
+            if media.screenVideo.state == .playable { return media.screenVideo.track }
+            if media.camera.state == .playable { return media.camera.track }
```
Comment (marcisme) [inline DailyKit/CallManager/Models/CallParticipant.swift]: Use `fit` for screenshare video.

File: DailyStarterKit/Layouts/GridLayoutView.swift [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+        init(manager: CallManageable) {
+            self.manager = manager
+            self.participants = manager.participants
+
+            manager.publisher(for: .participants)
+                .map { self.participants.stabilize($0) }
+                .assign(to: &$participants)
+        }
+
+        @Published private(set) var participants: CallParticipants
```
Comment (marcisme) [inline DailyStarterKit/Layouts/GridLayoutView.swift]: We should use `local` and `visible` here, so the view isn't invalidated when non-visible participants are updated.

File: DailyStarterKit/Call/CallControlsOverlayView.swift [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+    private var padding: EdgeInsets {
+        switch layout {
+        case .portrait:
+            return EdgeInsets(top: 16, leading: 16, bottom: 0, trailing: 16)
+        case .landscape:
+            // With default padding the top and bottom of the overlay do not stay within the safe area in
+            // landscape, so we increase the insets here.
+            return EdgeInsets(top: 30, leading: 16, bottom: 9, trailing: 16)
+        }
+    }
```
Comment (marcisme) [inline DailyStarterKit/Call/CallControlsOverlayView.swift:163]: @regexident This should handle the inset issue you noticed. Thanks!

File: DailyKit/CallManager/CallManager.swift (hunk 1 above)
Comment (regexident) [reply DailyKit/CallManager/CallManager.swift:43]: The individual subjects are [`CurrentValueSubject<T, E>`](https://developer.apple.com/documentation/combine/currentvaluesubject), i.e. streams that cache their most-recent received value.

The properties (getters) provide an ergonomic and synchronous access to individual "current" states (imperative programming), while the subjects allow for attaching event handlers to the updates (reactive programming). The properties aren't strictly necessary as the user could just as well access the subject's value themselves, but they make the code a bit cleaner and nicer to read.

A lot of the code in this starter kit is likely to get upstreamed into `daily-ios` at some point in some form or another. So in some semi-distant future the necessity for such user-provided wrappers (e.g. `CallManager`, `CallParticipants`, …) will hopefully go away entirely.

File: DailyKit/CallManager/Models/CallCamera.swift (hunk 3 above)
Comment (regexident) [reply DailyKit/CallManager/Models/CallCamera.swift]: +1!

It's an unofficial convention to put properties above initializers in Swift.

File: DailyKit/CallManager/Models/CallMuteState.swift (hunk 4 above)
Comment (regexident) [reply DailyKit/CallManager/Models/CallMuteState.swift:45]: It's a "catch all" wildcard.

So yeah, in other languages it would be called `default`.
Swift also has a `case default:` syntactic sugar for it, but `_` is more general.
(Personal pet peeve: the `case default:` should not exist, we already have `case _:` for that)

File: DailyKit/CallManager/Models/CallParticipants.swift (hunk 8 above)
Comment (regexident) [reply DailyKit/CallManager/Models/CallParticipants.swift]: Participants are value types, so updates to a participant's fields do not propagate between copies.

File: DailyKit/CallManager/Models/CallParticipants.swift (hunk 9 above)
Comment (regexident) [reply DailyKit/CallManager/Models/CallParticipants.swift:6]: +1!

Same as https://github.com/daily-demos/daily-ios-starter-kit/pull/1#discussion_r1255610183

File: DailyKitTests/CallParticipantsTests.swift (hunk 10 above)
Comment (regexident) [reply DailyKitTests/CallParticipantsTests.swift]: `// MARK: - Count` comments a special kind of comments that produces section headers and delimiters in Xcode's symbol dropdown, so keeping them brief is usually preferable:

image

File: DailyKitTests/CallParticipantsVisibleTests.swift (hunk 11 above)
Comment (regexident) [reply DailyKitTests/CallParticipantsVisibleTests.swift]: Yesss!!

File: DailyKit/CallManager/CallClientSubjects.swift [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+
+        setupPublishers()
+    }
+
+    private func setupPublishers() {
+        publishers[CallClientProperty.url] = url.eraseToAnyPublisher()
+        publishers[CallClientProperty.callState] = callState.eraseToAnyPublisher()
+        publishers[CallClientProperty.camera] = camera.eraseToAnyPublisher()
+        publishers[CallClientProperty.microphone] = microphone.eraseToAnyPublisher()
+        publishers[CallClientProperty.participants] = participants.eraseToAnyPublisher()
```
Comment (regexident) [inline DailyKit/CallManager/CallClientSubjects.swift:36]:
```suggestion
        publishers[.url] = url.eraseToAnyPublisher()
        publishers[.callState] = callState.eraseToAnyPublisher()
        publishers[.camera] = camera.eraseToAnyPublisher()
        publishers[.microphone] = microphone.eraseToAnyPublisher()
        publishers[.participants] = participants.eraseToAnyPublisher()
```

File: DailyKit/CallManager/CallClientSubjects.swift [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+        callState: CallState,
+        camera: CallCamera,
+        microphone: CallMicrophone,
+        participants: CallParticipants
+    ) {
+        self.url = CurrentValueSubject<URL?, Never>(url)
+        self.callState = CurrentValueSubject<CallState, Never>(callState)
+        self.camera = CurrentValueSubject<CallCamera, Never>(camera)
+        self.microphone = CurrentValueSubject<CallMicrophone, Never>(microphone)
+        self.participants = CurrentValueSubject<CallParticipants, Never>(participants)
```
Comment (regexident) [inline DailyKit/CallManager/CallClientSubjects.swift]: You shouldn't need the fully qualified names here, I think?

```suggestion
        self.url = .init(url)
        self.callState = .init(callState)
        self.camera = .init(camera)
        self.microphone = .init(microphone)
        self.participants = .init(participants)
```

File: DailyKit/CallManager/FakeCallManager.swift [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+        switch camera.mode {
+        case .user:
+            self.subjects.camera.send(CallCamera(video: camera.video, mode: .environment))
+        case .environment:
+            self.subjects.camera.send(CallCamera(video: camera.video, mode: .user))
+        }
+    }
+
+    public func setUsername(_ username: String?) {
+        guard let username = username else { fatalError("TODO") }
```
Comment (regexident) [inline DailyKit/CallManager/FakeCallManager.swift]: Unresolved todo.

File: DailyKit/CallManager/CallManager.swift (hunk 2 above)
Comment (marcisme) [reply DailyKit/CallManager/CallManager.swift:108]: Potentially long enough to need an explicit loading state. This is an optimization that I made in response to some feedback from someone testing the app in a location with poor network connectivity. I want to see if there's anything we can do in the SDK to update the state faster in those conditions. Ideally we can transition to `joining` immediately and remove this part, otherwise I think we'll want to introduce a UI-specific call state type that can represent both the `CallState` and user intent. `CallCallState`? 🙃

File: DailyKit/CallManager/Models/CallCamera.swift (hunk 3 above)
Comment (marcisme) [reply DailyKit/CallManager/Models/CallCamera.swift]: Properties above initializers isn't something I've seen be a hard style preference in the work I've done, but it is common. Personally I prefer more flexibility to order members within a given type. I think in many cases it can make initial reading and maintenance easier to have a property grouped next to the methods that depend on it, or to have some related members grouped together instead of spread around a file based on member type or visibility. 

Here's one example of the ["logical ordering"](https://google.github.io/swift/#type-variable-and-function-declarations) perspective.

> The order of types, variables, and functions in a source file, and the order of the members of those types, can have a great effect on readability. However, there is no single correct recipe for how to do it; different files and different types may order their contents in different ways.
>
> What is important is that each file and type uses some logical order, which its maintainer could explain if asked. For example, new methods are not just habitually added to the end of the type, as that would yield “chronological by date added” ordering, which is not a logical ordering.

In this case I think properties above the initializer is more logical because these are simple state types. I'll move it up here and in the other `Call*` structs. Thanks both of you for mentioning that.

File: DailyKit/CallManager/Models/CallMuteState.swift (hunk 13 above)
Comment (marcisme) [reply DailyKit/CallManager/Models/CallMuteState.swift:21]: Like other privacy settings, disabling camera or microphone authorization terminates the app in the background, so having these here is needed to ensure the state is correct. I've also added some guards to open `Settings.app` to the app-specific settings when authorization has been denied.

File: DailyKitTests/CallParticipantsVisibleTests.swift (hunk 11 above)
Comment (marcisme) [reply DailyKitTests/CallParticipantsVisibleTests.swift]: ❤️

File: DailyKit/CallManager/CallManager.swift (hunk 1 above)
Comment (marcisme) [reply DailyKit/CallManager/CallManager.swift:43]: > Is its purpose just to hold the publishers collection?

A common type for the publishers is the main purpose. I wanted to avoid duplicating this code in both implementations of `CallManageable`, especially with the casting / `fatalError` part. As @regexident mentioned, we're thinking about adding something like `func publisher<T>(for property: CallClientProperty<T>) -> AnyPublisher<T, Never>` to the `CallClient`, so this was also an effort to start thinking through a general abstraction we might want to use there too.

File: DailyKit/CallManager/Models/CallParticipant.swift (hunk 7 above)
Comment (marcisme) [reply DailyKit/CallManager/Models/CallParticipant.swift:61]: I think it's the session ID, but I haven't seen it referred to that anywhere in the code. We have an optional user ID available at [participant.info.userID](https://reference-ios.daily.co/documentation/daily/participantinfo).

File: DailyKit/CallManager/Models/CallParticipant.swift (hunk 6 above)
Comment (marcisme) [reply DailyKit/CallManager/Models/CallParticipant.swift]: Personally I prefer not destructing types like that most of the time. I've worked on many projects where consistent naming isn't followed, which then makes using search and replace for refactoring really hard because you can't search for consistent property names. You have to find the property, figure out what it was renamed to as a local, then search for that, and you have to do that at every call site. Xcode doesn't have great refactoring support compared to some other editors, so that probably influences my perspective.

I don't mind changing it, but since this is a nit, I'd prefer to not introduce locals here.

File: DailyKit/CallManager/Models/CallParticipants.swift (hunk 9 above)
Comment (marcisme) [reply DailyKit/CallManager/Models/CallParticipants.swift:6]: See my [comment on #1](https://github.com/daily-demos/daily-ios-starter-kit/pull/1#discussion_r1256522531) too. I wanted to have the `Builder` type above the properties for `CallParticipants` itself. It's gotten a lot bigger though, so I'll move the `Builder` to its own file.

File: DailyKitTests/CallParticipantsTests.swift (hunk 10 above)
Comment (marcisme) [reply DailyKitTests/CallParticipantsTests.swift]: Related to the comment above, I moved the `CallParticipants.Builder` to its own file and consolidated the test files to one for each of the two types. I also made the `MARK`s a little more descriptive.

File: DailyKit/CallManager/CallClientSubjects.swift (hunk 17 above)
Comment (marcisme) [reply DailyKit/CallManager/CallClientSubjects.swift:36]: Unfortunately that doesn't work because the key for the dictionary is `AnyHashable`. With `CallClientProperty` being generic, I don't think there's a way to use it as a key directly. I'd be interested to know if there is though.

File: DailyKit/CallManager/FakeCallManager.swift (hunk 19 above)
Comment (marcisme) [reply DailyKit/CallManager/FakeCallManager.swift]: Thanks!

File: DailyKit/CallManager/Models/CallCamera.swift (hunk 3 above)
Comment (regexident) [reply DailyKit/CallManager/Models/CallCamera.swift]: Afaict it's by far the most prevalent unspoken community convention.

Even Apple uses it pretty consistently:
https://docs.swift.org/swift-book/documentation/the-swift-programming-language/initialization/

File: DailyKit/CallManager/CallClientSubjects.swift (hunk 17 above)
Comment (regexident) [reply DailyKit/CallManager/CallClientSubjects.swift:36]: Ah, gotcha!

File: .gitignore [hunk 20]
```diff
@@ -0,0 +1 @@
+xcuserdata/
```
Comment (regexident) [inline .gitignore:43]: This should probably also include `.DS_Store` and the like?

I usually go with the [gitignore.io](https://gitignore.io) template for [macos + swift](https://www.toptal.com/developers/gitignore/api/macos,swift).

**Review by regexident — APPROVED** [review]
Tiny nit about `.gitignore`, otherwise LGTM!

**Review by regexident — APPROVED** [review]
LGTM
